plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformer.kt

```diff
@@ -0,0 +1,595 @@
+                    val elementIndex = getValueArgument(0)!!.deepCopyWithVariables()
+                    buildCall(
+                        target = getSymbol,
+                        type = valueType,
+                        origin = IrStatementOrigin.LAMBDA,
+                        valueArguments = listOf(elementIndex)
+                    ).apply {
+                        dispatchReceiver = getField.deepCopyWithVariables()
+                    }
+                } else {
+                    getField.deepCopyWithVariables()
+                }
+            }
+            this.body = buildBlockBody(listOf(body))
+            origin = IrDeclarationOrigin.DEFAULT_PROPERTY_ACCESSOR
+        }
+        return IrFunctionExpressionImpl(
+            UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+            type,
+            accessorFunction,
+            IrStatementOrigin.LAMBDA
+        )
+    }
+
+    private fun buildSetField(backingField: IrField, ownerClass: IrExpression?, value: IrGetValue): IrSetField {
+        val receiver = if (ownerClass is IrTypeOperatorCall) ownerClass.argument as IrGetValue else ownerClass
+        val fieldSymbol = backingField.symbol
+        return buildSetField(
+            symbol = fieldSymbol,
+            receiver = receiver,
+            value = value
+        )
+    }
+
+    private fun buildGetField(backingField: IrField, ownerClass: IrExpression?): IrGetField {
+        val receiver = if (ownerClass is IrTypeOperatorCall) ownerClass.argument as IrGetValue else ownerClass
+        return buildGetField(
+            symbol = backingField.symbol,
+            receiver = receiver
+        )
+    }
+
+    private fun IrCall.getAccessors(parentDeclaration: IrDeclaration): List<IrExpression> =
+        listOf(buildAccessorLambda(isSetter = false, parentDeclaration = parentDeclaration),
+               buildAccessorLambda(isSetter = true, parentDeclaration = parentDeclaration))
+
+    private fun getRuntimeFunctionSymbol(name: String, type: IrType): IrSimpleFunctionSymbol {
+        val functionName = when (name) {
+            "value.<get-value>" -> "getValue"
+            "value.<set-value>" -> "setValue"
+            else -> name
+        }
+        return referencePackageFunction("kotlinx.atomicfu", "$ATOMICFU_RUNTIME_FUNCTION_PREDICATE$functionName") {
+            val typeArg = it.owner.getGetterReturnType()
+            !(typeArg as IrType).isPrimitiveType() || typeArg == type
+        }
+    }
+
+    private fun IrFunction.getGetterReturnType() = (valueParameters[valueParameters.lastIndex - 1].type as IrSimpleType).arguments.first().typeOrNull
+
+    private fun IrCall.isAtomicFactory(): Boolean {
+        val name = symbol.owner.name
+        return !name.isSpecial && name.identifier == ATOMIC_CONSTRUCTOR
+    }
+
+    private fun IrCall.isAtomicArrayFactory(): Boolean {
+        val name = symbol.owner.name
+        return !name.isSpecial && name.identifier == ATOMIC_ARRAY_FACTORY_FUNCTION
```

Comment:
`name.asString() == ATOMIC_ARRAY_FACTORY_FUNCTION`

Reply:
Or even better, make `ATOMIC_CONSTRUCTOR` `Name` and use simple `name == ATOMIC_CONSTRUCTOR`